### PR TITLE
Initial typechecking task/ARXIVNG-285

### PR DESCRIPTION
**Pull request by JaimieMurdock**
An initial round of type checking. Doesn't yet capture completely missing annotations, but gets to the default settings of `mypy --ignore-missing-imports`. Will open a new issue to deal with those, but wanted to merge early and often!

### CODE REVIEW

File: search/controllers/simple/__init__.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def search(request_params: dict) -> Response:
 
     # Fall back to form-based search.
     form = SimpleSearchForm(request_params)
-    q = None
+    q: Query
```
Comment (erickpeirson) [inline search/controllers/simple/__init__.py]: See line 98: we do actually set this to None; should it be ``Optional[Query]``?

File: search/domain/base.py [hunk 2]
```diff
@@ -83,11 +83,8 @@ class SchemaBase(Base):
 
     def __getattr__(self, key: str) -> Any:
         """Get a schema attribute."""
-        try:
-            super(SchemaBase, self).__getattr__(key)
```
Comment (erickpeirson) [inline search/domain/base.py]: Turns out ``dict`` has no ``__getattr__`` (of course...)

Comment (JaimieMurdock) [reply search/domain/base.py]: That's so strange.

File: search/controllers/simple/__init__.py (hunk 1 above)
Comment (JaimieMurdock) [reply search/controllers/simple/__init__.py]: Here's some fun:

```
>>> a: int
>>> a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'a' is not defined
>>> a = None
>>> a
>>> 
```

submitting a patch for `Optional[Query]` then merging.

Comment (JaimieMurdock) [reply search/controllers/simple/__init__.py]: Patch will be on all controllers with that type of error.

**Pull request merged**